File: src/schedule.js
// Turns raw LMS calendar entries into the day/week view the UI renders.
//
// Every class shows up in the calendar twice: once as a session event (batch-wide, e.g.
// "Maths - Section B - Session 2") and once as an attendance event (personalised, e.g.
// "Attendance - Maths - Section B - Session 2") that carries the nid you mark against.
// We merge the pair into one row, and filter the leftover batch-wide events down to
// your own section and group.

const COHORT_RE = /\b(Section\s+[A-Z](?:\s*&\s*[A-Z])*|Group\s+\d+)\b/i;

export function parseLmsTime(value) {
  const m = /^(\d{4})-(\d{2})-(\d{2})[ T](\d{2}):(\d{2})(?::(\d{2}))?/.exec(String(value || ''));
  if (!m) return null;
  // LMS timestamps are wall-clock IST with no offset; the machine running this is on IST too,
  // so constructing a local Date is the correct reading.
  return new Date(+m[1], +m[2] - 1, +m[3], +m[4], +m[5], +(m[6] || 0));
}

export function ymd(date) {
  const p = (n) => String(n).padStart(2, '0');
  return `${date.getFullYear()}-${p(date.getMonth() + 1)}-${p(date.getDate())}`;
}

export function addDays(date, days) {
  const next = new Date(date);
  next.setDate(next.getDate() + days);
  return next;
}

/** "Attendance - Maths - Section B - Session 2" -> {name:"Maths", cohort:"Section B", session:2} */
export function parseTitle(rawTitle) {
  let rest = String(rawTitle || '').replace(/^\s*Attendance\s*-\s*/i, '').trim();
  let session = null;
  let cohort = null;

  const sessionMatch = rest.match(/\s*-\s*Session\s+(\d+)\s*$/i);
  if (sessionMatch) {
    session = Number(sessionMatch[1]);
    rest = rest.slice(0, sessionMatch.index).trim();
  }
  const cohortMatch = rest.match(/\s*-\s*(Section\s+[A-Z](?:\s*&\s*[A-Z])*|Group\s+\d+)\s*$/i);
  if (cohortMatch) {
    cohort = cohortMatch[1].replace(/\s+/g, ' ').trim();
    rest = rest.slice(0, cohortMatch.index).trim();
  }
  return { name: rest || String(rawTitle || '').trim(), cohort, session };
}

function mergeKey(event) {
  const { name, cohort, session } = parseTitle(event.title);
  return `${name.toLowerCase()}|${(cohort || '').toLowerCase()}|${session ?? ''}|${event.start}`;
}

const isAttendanceEvent = (event) => typeof event.url === 'string' && event.url.includes('/classroom/');

/**
 * Work out which section and groups you belong to, by looking at the attendance sessions
 * the LMS personalises to you. "Section A & B" is shared with the other section, so only
 * single-cohort tokens are treated as evidence.
 */
export function detectCohorts(events) {
  const sections = new Set();
  const groups = new Set();
  for (const event of events) {
    if (!isAttendanceEvent(event)) continue;
    const { cohort } = parseTitle(event.title);
    if (!cohort) continue;
    const group = cohort.match(/^Group\s+(\d+)$/i);
    if (group) { groups.add(group[1]); continue; }
    const letters = cohort.match(/\b[A-Z]\b/g) || [];
    if (letters.length === 1) sections.add(letters[0].toUpperCase());
  }
  return { sections, groups };
}

export function parseCohortOverride(text) {
  const sections = new Set();
  const groups = new Set();
  for (const token of String(text || '').split(',').map((t) => t.trim()).filter(Boolean)) {
    const group = token.match(/^Group\s+(\d+)$/i);
    if (group) { groups.add(group[1]); continue; }
    for (const letter of token.match(/\b[A-Z]\b/g) || []) sections.add(letter.toUpperCase());
  }
  return sections.size || groups.size ? { sections, groups } : null;
}

/** Does this batch-wide event belong to you? Events with no cohort marker are for everyone. */
function belongsToMe(event, cohorts) {
  const token = (event.title.match(COHORT_RE) || [])[1];
  if (!token) return true;
  const group = token.match(/^Group\s+(\d+)$/i);
  if (group) return cohorts.groups.size === 0 || cohorts.groups.has(group[1]);
  const letters = (token.match(/\b[A-Z]\b/g) || []).map((l) => l.toUpperCase());
  if (!letters.length) return true;
  if (cohorts.sections.size === 0) return true;
  return letters.some((letter) => cohorts.sections.has(letter));
}

/**
 * Merge raw calendar events into one row per class.
 * Rows carry `nid` when the class has an attendance session you can mark.
 */
export function buildSessions(events, cohorts) {
  const byKey = new Map();

  for (const event of events) {
    const key = mergeKey(event);
    const start = parseLmsTime(event.start);
    const end = parseLmsTime(event.end);
    if (!start) continue;

    let row = byKey.get(key);
    if (!row) {
      const { name, cohort, session } = parseTitle(event.title);
      row = {
        key,
        name,
        cohort,
        session,
        start: start.toISOString(),
        end: end ? end.toISOString() : null,
        startMs: start.getTime(),
        endMs: end ? end.getTime() : start.getTime(),
        nid: null,
        eventNid: null,
        subject: null,
        trainer: null,
        room: null,
        mine: true,
      };
      byKey.set(key, row);
    }

    if (isAttendanceEvent(event)) {
      row.nid = String(event.nid);
      row.subject = event.subject || row.subject;
      row.trainer = event.trainers || row.trainer;
    } else {
      row.eventNid = String(event.nid);
      // An attendance row proves the class is yours regardless of how the title reads;
      // otherwise fall back to matching the cohort in the title.
      row.mine = row.nid ? true : belongsToMe(event, cohorts);
    }
  }

  return [...byKey.values()]
    .filter((row) => row.nid || row.mine)
    .sort((a, b) => a.startMs - b.startMs || a.name.localeCompare(b.name));
}

/**
 * The UI state for a session.
 *   marked   - already recorded
 *   open     - the LMS will accept a mark right now
 *   upcoming - hasn't opened yet
 *   missed   - window has passed and nothing was recorded
 *   closed   - nothing to mark (no attendance session attached)
 */
export function sessionState(row, now) {
  if (!row.nid) return row.endMs < now ? 'done' : 'noattendance';
  if (row.marked) return 'marked';
  if (row.markable) return 'open';
  if (now < row.startMs) return 'upcoming';
  return 'missed';
}

File: android/app/src/main/java/org/isdm/companion/domain/Schedule.kt
package org.isdm.companion.domain

import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.LinkedHashMap
import java.util.Locale

/** The LMS sends wall-clock timestamps in IST without an offset. */
val LMS_ZONE: ZoneId = ZoneId.of("Asia/Kolkata")

private val LMS_TIME_RE = Regex(
    """^(\d{4})-(\d{2})-(\d{2})[ T](\d{2}):(\d{2})(?::(\d{2}))?""",
)
private val ATTENDANCE_PREFIX_RE = Regex("^\\s*Attendance\\s*-\\s*", RegexOption.IGNORE_CASE)
private val SESSION_SUFFIX_RE = Regex("\\s*-\\s*Session\\s+(\\d+)\\s*$", RegexOption.IGNORE_CASE)
private val COHORT_SUFFIX_RE = Regex(
    "\\s*-\\s*(Section\\s+[A-Z](?:\\s*&\\s*[A-Z])*|Group\\s+\\d+)\\s*$",
    RegexOption.IGNORE_CASE,
)
private val COHORT_RE = Regex(
    "\\b(Section\\s+[A-Z](?:\\s*&\\s*[A-Z])*|Group\\s+\\d+)\\b",
    RegexOption.IGNORE_CASE,
)
private val GROUP_RE = Regex("^Group\\s+(\\d+)$", RegexOption.IGNORE_CASE)
private val SECTION_LETTER_RE = Regex("\\b[A-Z]\\b")

/**
 * Parse the LMS's wall-clock timestamp into an Instant in [LMS_ZONE].
 *
 * The JavaScript implementation accepts a prefix and ignores trailing text.  It also lets
 * Date normalise out-of-range month/day/hour values.  Building from the first day of the year
 * and applying offsets preserves those two useful behaviours while keeping the zone explicit.
 */
fun parseLmsTime(value: String?): Instant? {
    val match = LMS_TIME_RE.find(value.orEmpty()) ?: return null
    val year = match.groupValues[1].toIntOrNull() ?: return null
    val month = match.groupValues[2].toLongOrNull() ?: return null
    val day = match.groupValues[3].toLongOrNull() ?: return null
    val hour = match.groupValues[4].toLongOrNull() ?: return null
    val minute = match.groupValues[5].toLongOrNull() ?: return null
    val second = match.groupValues.getOrNull(6)?.toLongOrNull() ?: 0L

    return try {
        LocalDateTime.of(year, 1, 1, 0, 0)
            .plusMonths(month - 1L)
            .plusDays(day - 1L)
            .plusHours(hour)
            .plusMinutes(minute)
            .plusSeconds(second)
            .atZone(LMS_ZONE)
            .toInstant()
    } catch (_: RuntimeException) {
        null
    }
}

fun ymd(date: LocalDate): String = date.format(DateTimeFormatter.ISO_LOCAL_DATE)

fun ymd(instant: Instant, zone: ZoneId = LMS_ZONE): String = ymd(instant.atZone(zone).toLocalDate())

fun addDays(date: LocalDate, days: Long): LocalDate = date.plusDays(days)

/** Parse "Attendance - Maths - Section B - Session 2" into its display pieces. */
fun parseTitle(rawTitle: String?): ParsedTitle {
    val original = rawTitle.orEmpty()
    var rest = original.replaceFirst(ATTENDANCE_PREFIX_RE, "").trim()
    var session: Int? = null
    var cohort: String? = null

    val sessionMatch = SESSION_SUFFIX_RE.find(rest)
    if (sessionMatch != null) {
        session = sessionMatch.groupValues[1].toIntOrNull()
        rest = rest.substring(0, sessionMatch.range.first).trim()
    }

    val cohortMatch = COHORT_SUFFIX_RE.find(rest)
    if (cohortMatch != null) {
        cohort = cohortMatch.groupValues[1].replace(Regex("\\s+"), " ").trim()
        rest = rest.substring(0, cohortMatch.range.first).trim()
    }

    return ParsedTitle(
        name = rest.ifEmpty { original.trim() },
        cohort = cohort,
        session = session,
    )
}

/** Is this the personalised attendance row rather than the batch-wide event row? */
fun isAttendanceEvent(event: CalendarEvent): Boolean = event.url?.contains("/classroom/") == true

/** Infer section/group membership from the LMS's personalised attendance rows. */
fun detectCohorts(events: Iterable<CalendarEvent>): Cohorts {
    val sections = linkedSetOf<String>()
    val groups = linkedSetOf<String>()

    for (event in events) {
        if (!isAttendanceEvent(event)) continue
        val cohort = parseTitle(event.title).cohort ?: continue
        val group = GROUP_RE.matchEntire(cohort)
        if (group != null) {
            groups += group.groupValues[1]
            continue
        }
        val letters = SECTION_LETTER_RE.findAll(cohort).map { it.value }.toList()
        if (letters.size == 1) sections += letters.single().uppercase(Locale.ROOT)
    }

    return Cohorts(sections = sections, groups = groups)
}

/** Parse the comma-separated COHORTS override used by the desktop setup. */
fun parseCohortOverride(text: String?): Cohorts? {
    val sections = linkedSetOf<String>()
    val groups = linkedSetOf<String>()

    for (token in text.orEmpty().split(',').map { it.trim() }.filter { it.isNotEmpty() }) {
        val group = GROUP_RE.matchEntire(token)
        if (group != null) {
            groups += group.groupValues[1]
            continue
        }
        SECTION_LETTER_RE.findAll(token)
            .map { it.value.uppercase(Locale.ROOT) }
            .forEach { sections += it }
    }

    return if (sections.isNotEmpty() || groups.isNotEmpty()) {
        Cohorts(sections = sections, groups = groups)
    } else {
        null
    }
}

/** Does a batch-wide event belong to the caller's section/group? */
private fun belongsToMe(event: CalendarEvent, cohorts: Cohorts): Boolean {
    val token = COHORT_RE.find(event.title)?.groupValues?.getOrNull(1) ?: return true
    val group = GROUP_RE.matchEntire(token)
    if (group != null) return cohorts.groups.isEmpty() || group.groupValues[1] in cohorts.groups

    val letters = SECTION_LETTER_RE.findAll(token).map { it.value.uppercase(Locale.ROOT) }.toList()
    if (letters.isEmpty()) return true
    if (cohorts.sections.isEmpty()) return true
    return letters.any { it in cohorts.sections }
}

private data class BuildingSession(
    val key: String,
    val name: String,
    val cohort: String?,
    val session: Int?,
    val start: Instant,
    val end: Instant?,
    var nid: String? = null,
    var eventNid: String? = null,
    var subject: String? = null,
    var trainer: String? = null,
    var mine: Boolean = true,
)

private fun mergeKey(event: CalendarEvent): String {
    val title = parseTitle(event.title)
    return "${title.name.lowerRoot()}|${title.cohort?.lowerRoot().orEmpty()}|${title.session ?: ""}|${event.start}"
}

/**
 * Merge duplicate batch-wide/attendance entries into one row and filter rows outside the
 * caller's section/group.  The returned order is start time, then name, like the JS client.
 */
fun buildSessions(events: Iterable<CalendarEvent>, cohorts: Cohorts): List<Session> {
    val byKey = LinkedHashMap<String, BuildingSession>()

    for (event in events) {
        val start = parseLmsTime(event.start) ?: continue
        val end = parseLmsTime(event.end)
        val key = mergeKey(event)
        val title = parseTitle(event.title)
        val row = byKey.getOrPut(key) {
            BuildingSession(
                key = key,
                name = title.name,
                cohort = title.cohort,
                session = title.session,
                start = start,
                end = end,
            )
        }

        if (isAttendanceEvent(event)) {
            row.nid = event.nid
            if (!event.subject.isNullOrEmpty()) row.subject = event.subject
            if (!event.trainers.isNullOrEmpty()) row.trainer = event.trainers
        } else {
            row.eventNid = event.nid
            // An attendance row proves ownership regardless of the title.  This is also
            // faithful to the JS implementation when the batch-wide row arrives first.
            row.mine = if (!row.nid.isNullOrEmpty()) true else belongsToMe(event, cohorts)
        }
    }

    return byKey.values
        .asSequence()
        .filter { !it.nid.isNullOrEmpty() || it.mine }
        .map {
            Session(
                name = it.name,
                cohort = it.cohort,
                session = it.session,
                start = it.start,
                end = it.end,
                nid = it.nid,
                eventNid = it.eventNid,
                subject = it.subject,
                trainer = it.trainer,
            )
        }
        .sortedWith(compareBy<Session> { it.startMs }.thenBy { it.name })
        .toList()
}

/** Compute the same six UI states emitted by the desktop service. */
fun sessionState(row: Session, now: Instant): SessionState {
    if (row.nid.isNullOrEmpty()) {
        return if (row.endMs < now.toEpochMilli()) SessionState.DONE else SessionState.NO_ATTENDANCE
    }
    if (row.marked) return SessionState.MARKED
    if (row.markable) return SessionState.OPEN
    if (now.toEpochMilli() < row.startMs) return SessionState.UPCOMING
    return SessionState.MISSED
}

fun sessionState(row: Session, nowMs: Long): SessionState = sessionState(row, Instant.ofEpochMilli(nowMs))
